proof/refine/ARM/KHeap_R.thy

```diff
@@ -3921,4 +3927,259 @@ lemma threadGet_getObject:
+  apply (clarsimp simp: projectKO_reply)
+  done
+
+lemma invs'_tcbs_scs_sym_refs:
+  "\<lbrakk>sym_refs (state_refs_of' s); pspace_aligned' s; pspace_distinct' s; valid_objs' s\<rbrakk>
+   \<Longrightarrow> tcbs_scs_sym_refs s"
+  apply (clarsimp simp: sym_heap_def)
+  apply (intro iffI conjI impI allI)
+   apply (frule (2) TCBSC_pred_map_state_refs_of')
+   apply (frule (1) sym_refsD, simp add: SCTcb_state_refs_of' obj_at'_sc_tcbs_of_equiv)
+  apply (frule (2) SCTcb_pred_map_state_refs_of')
+   apply (frule (1) sym_refsD, simp add: TCBSchedContext_state_refs_of' obj_at'_tcb_scs_of_equiv)
+  done
+
+lemma sym_refs_replies_scs:
+  "\<lbrakk>sym_refs (state_refs_of' s); pspace_aligned' s; pspace_distinct' s; valid_objs' s\<rbrakk>
+   \<Longrightarrow> replies_scs_sym_refs s"
+  apply (clarsimp simp: sym_heap_def)
+  apply (intro iffI conjI impI allI)
+   apply (frule (2) SCReply_pred_map_state_refs_of')
+   apply (frule (1) sym_refsD, simp add: ReplySchedContext_state_refs_of' obj_at'_replySCs_of_equiv)
+  apply (frule (2) ReplySC_pred_map_state_refs_of')
+   apply (frule (1) sym_refsD, simp add: SCReply_state_refs_of' obj_at'_scReplies_of_equiv)
+  done
+
+lemma setSchedContext_scTCBs_of:
+  "\<lbrace>\<lambda>s. P (\<lambda>a. if a = scPtr then scTCB sc else scTCBs_of s a)\<rbrace>
+   setSchedContext scPtr sc
+   \<lbrace>\<lambda>_ s. P (scTCBs_of s)\<rbrace>"
+  unfolding setSchedContext_def
+  apply (wpsimp wp: setObject_sc_wp)
+  apply (erule back_subst[where P=P], rule ext)
+  by (clarsimp simp: opt_map_def)
+
+lemma setSchedContext_scReplies_of:
+  "\<lbrace>\<lambda>s. P (\<lambda>a. if a = scPtr then scReply sc else scReplies_of s a)\<rbrace>
+   setSchedContext scPtr sc
+   \<lbrace>\<lambda>_ s. P (scReplies_of s)\<rbrace>"
+  unfolding setSchedContext_def
+  apply (wpsimp wp: setObject_sc_wp)
+  apply (erule back_subst[where P=P], rule ext)
+  by (clarsimp simp: opt_map_def)
+
+lemma getObject_tcb_wp:
+  "\<lbrace>\<lambda>s. tcb_at' p s \<longrightarrow> (\<exists>t::tcb. ko_at' t p s \<and> Q t s)\<rbrace> getObject p \<lbrace>Q\<rbrace>"
+  by (clarsimp simp: getObject_def valid_def in_monad
+                     split_def objBits_simps' loadObject_default_def
+                     projectKOs obj_at'_def in_magnitude_check)
+
+lemma threadSet_tcb_scs_of':
+  "\<lbrace>\<lambda>s. P (\<lambda>a. if a = t then tcbSchedContext (f (the (tcbs_of' s a))) else tcb_scs_of' s a)\<rbrace>
+   threadSet f t
+   \<lbrace>\<lambda>_ s. P (tcb_scs_of' s)\<rbrace>"
+  unfolding threadSet_def
+  apply (wpsimp wp: setObject_tcb_wp getObject_tcb_wp)
+  apply (clarsimp simp: tcb_at'_ex_eq_all)
+  apply (erule back_subst[where P=P], rule ext)
+  apply (clarsimp simp: opt_map_def obj_at'_real_def ko_wp_at'_def projectKOs)
+  done
+
+lemma replyNexts_Some_replySCs_None:
+  "replyNexts_of s rPtr \<noteq> None \<Longrightarrow> replySCs_of s rPtr = None"
+   by (clarsimp simp: obj_at'_real_def ko_wp_at'_def projectKOs opt_map_def split: option.split)
```

Comment:
Something very similar (and have the same name) exists in Reply_R, and that one is in a pair with the opposite. This one is actually a better statement. Can you please add the opposite version here, too? And maybe replace the one in Reply_R with this one?